### PR TITLE
optimism: use op-alloy TxDeposit

### LINKED ISSUE #9484: use alloy transaction types
the core tx types are quivalent to alloy's:

https://github.com/paradigmxyz/reth/blob/67f617646ce4b68a5096239238aa9209636ff500/crates/primitives/src/transaction/eip1559.rs#L13-L13

the only difference is the codec impl

https://github.com/paradigmxyz/reth/blob/67f617646ce4b68a5096239238aa9209636ff500/crates/primitives/src/transaction/eip1559.rs#L11-L11

we can implement codec by doing the same as for

https://github.com/paradigmxyz/reth/blob/67f617646ce4b68a5096239238aa9209636ff500/crates/storage/codecs/src/alloy/withdrawal.rs#L6-L11

## TODO
- https://github.com/paradigmxyz/reth/pull/9593
- https://github.com/paradigmxyz/reth/pull/10262
- https://github.com/paradigmxyz/reth/pull/10623
- https://github.com/paradigmxyz/reth/pull/10624
- https://github.com/paradigmxyz/reth/pull/10617
- [ ] optimism deposit

This should be done in separate prs and involves
* move existing struct def to https://github.com/paradigmxyz/reth/blob/67f617646ce4b68a5096239238aa9209636ff500/crates/storage/codecs/src/alloy/transaction.rs#L1-L1 impl codec for alloy types via into/from
* phase out primitive transaction defs

there are likely missing/renamed functions in alloy, so we need to surface those and add to alloy if appropriate

### PR DESCRIPTION
**Pull request by tcoratger**
Should close https://github.com/paradigmxyz/reth/issues/9484

Waiting for the following inclusion in next op-alloy release:
- https://github.com/alloy-rs/op-alloy/pull/68
- https://github.com/alloy-rs/op-alloy/pull/69

### CODE REVIEW

**Comment by mattsse** [comment]
new op-alloy patch out

File: crates/storage/codecs/src/alloy/transaction/mod.rs
[the remark this answers is not in the document]
Comment (tcoratger) [reply crates/storage/codecs/src/alloy/transaction/mod.rs:1]: done

File: crates/storage/codecs/src/alloy/transaction/mod.rs [hunk 1]
```diff
@@ -13,6 +15,8 @@ mod tests {
     // this check is to ensure we do not inadvertently add too many fields to a struct which would
     // expand the flags field and break backwards compatibility
 
+    #[cfg(feature = "optimism")]
+    use super::optimism::TxDeposit;
```
Comment (mattsse) [inline crates/storage/codecs/src/alloy/transaction/mod.rs:19]: pedantic nit, use full paths instead

File: crates/storage/codecs/src/lib.rs [hunk 2]
```diff
@@ -21,6 +21,8 @@ pub use reth_codecs_derive::*;
 
 use alloy_primitives::{Address, Bloom, Bytes, FixedBytes, U256};
 use bytes::{Buf, BufMut};
+#[cfg(feature = "optimism")]
+use op_alloy_consensus as _;
```
Comment (mattsse) [inline crates/storage/codecs/src/lib.rs:25]: why is this necessary?

File: crates/primitives/src/lib.rs [hunk 3]
```diff
@@ -63,6 +63,8 @@ pub use transaction::{
     PooledTransactionsElementEcRecovered,
 };
 
+#[cfg(feature = "optimism")]
+use reth_codecs as _;
```
Comment (mattsse) [inline crates/primitives/src/lib.rs:67]: why is this now necessary?

**Review by mattsse — COMMENTED** [review]
last questions

pending @klkvr

File: crates/storage/codecs/src/alloy/transaction/optimism.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        let tx = TxDeposit {
+            source_hash: self.source_hash,
+            from: self.from,
+            to: self.to,
+            mint: self.mint,
+            value: self.value,
+            gas_limit: self.gas_limit as u64,
+            is_system_transaction: self.is_system_transaction,
+            input: self.input.clone(),
```
Comment (klkvr) [inline crates/storage/codecs/src/alloy/transaction/optimism.rs:42]: @mattsse currently all `Compact` impls for alloy transactions clone input during encoding. should we address this with a helper struct?

**Review by klkvr — APPROVED** [review]
lgtm, pending @mattsse nits

File: crates/storage/codecs/src/alloy/transaction/optimism.rs (hunk 4 above)
Comment (mattsse) [reply crates/storage/codecs/src/alloy/transaction/optimism.rs:42]: I think this is fine because this is `Bytes` and should be cheap enough

File: crates/storage/codecs/src/lib.rs (hunk 2 above)
Comment (tcoratger) [reply crates/storage/codecs/src/lib.rs:25]: Maybe I missed something with the imports but otherwise it triggers the `unused_crate_dependencies` clippy (maybe a false positive). I have the impression that since `op_alloy_consensus` is only used in a submodule under the `optimism` cfg, it creates this weird warning...

However I thought I did things right in the `Cargo.toml` with an optional and a dep only for the `optimism` cfg..

File: crates/primitives/src/lib.rs (hunk 3 above)
Comment (tcoratger) [reply crates/primitives/src/lib.rs:67]: Same reason...

File: crates/storage/codecs/src/lib.rs (hunk 2 above)
Comment (tcoratger) [reply crates/storage/codecs/src/lib.rs:25]: Imo this is similar to what we have here 

https://github.com/paradigmxyz/reth/blob/5df03fb3c3d70b3974ede5da7174bd9b9336ded0/crates/primitives/src/alloy_compat.rs#L11
